pkg/workerpool/workerpool.go

```diff
@@ -0,0 +1,130 @@
+package workerpool
+
+import (
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/openservicemesh/osm/pkg/logger"
+)
+
+const (
+	// Size of the job queue per worker
+	maxJobPerWorker = 4096
+)
+
+var (
+	log = logger.New("workerpool")
+)
+
+// Worker context for a worker routine
+type Worker struct {
+	id            int
+	jobs          chan Job        // Job queue
+	stop          chan struct{}   // Stop channel
+	wg            *sync.WaitGroup // Pointer to WorkerPool wg
+	JobsProcessed uint64          // Jobs processed by this worker
+}
+
+// WorkerPool object representation
+type WorkerPool struct {
+	wg            sync.WaitGroup // Sync group, to stop workers if needed
+	workerContext []*Worker      // Worker contexts
+	nWorkers      uint64         // Number of workers. Uint64 for easier mod hash later
+	rRobinCounter uint64         // Used only by the round robin api. Modified atomically on API.
+}
+
+// Job is a runnable interface to queue jobs on a WorkerPool
+type Job interface {
+	JobName() string // Returns name for a job
+	Hash() uint64    // Returns a uint64 hash for a job.
+	Run()            // Executes the job
+}
+
+// NewWorkerPool creates a new work group.
+// If nWorkers is 0, will poll goMaxProcs to get the number of routines to spawn.
+// Reminder routines are never pinned to system threads, it's up to the go scheduler to decide
+// where these will be run.
+func NewWorkerPool(nWorkers int) *WorkerPool {
+	if nWorkers == 0 {
+		// read GOMAXPROCS, -1 to avoid changing it
+		nWorkers = runtime.GOMAXPROCS(-1)
+	}
+
+	log.Info().Msgf("New worker pool setting up %d workers", nWorkers)
+
+	workPool := WorkerPool{
+		wg:            sync.WaitGroup{},
+		workerContext: []*Worker{},
+		rRobinCounter: 0,
+		nWorkers:      0,
+	}
```

Comment:
`var workpool WorkerPool` doesn't suffice?

Reply:
oh. I thought slices in a struct require init (either copy/move, or `make()`)

Reply:
Since its a slice, it will allocate on append if current capacity is not enough. I'd be surprised if default initialization doesn't work.